### ISSUE TITLE
Cant load model after adding custom component in pipeline

### ISSUE THREAD

**Issue by duesXMachine** [explosion/spaCy#1984]
Hi, I trained a NER from scratch with custom dataset without any other component in pipeline. I loaded the model and added another component(Phrase Matcher) to existing pipeline  and trained NER again.Now I cant load the new saved model.
```nlp = spacy.load('new_model')
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/home/deusxmachine/.local/lib/python2.7/site-packages/spacy/__init__.py", line 19, in load
    return util.load_model(name, **overrides)
  File "/home/deusxmachine/.local/lib/python2.7/site-packages/spacy/util.py", line 116, in load_model
    return load_model_from_path(Path(name), **overrides)
  File "/home/deusxmachine/.local/lib/python2.7/site-packages/spacy/util.py", line 156, in load_model_from_path
    component = nlp.create_pipe(name, config=config)
  File "/home/deusxmachine/.local/lib/python2.7/site-packages/spacy/language.py", line 215, in create_pipe
    raise KeyError("Can't find factory for '{}'.".format(name))
KeyError: u"Can't find factory for 'my_component'."
```

## Your Environment

* **Operating System**: Linux-4.13.0-32-generic-x86_64-with-Ubuntu-17.10-artful
* **Python Version Used**: 2.7.14
* **spaCy Version Used**: 2.0.7

**Comment by thetravelingsalesman** [comment]
Also having the same issue with default en_core_web_lg/sm.

**Comment by ines** [comment]
Sorry for only getting to this now. The reason you're seeing this error is because spaCy doesn't have a built-in factory for your component. When the model is loaded, spaCy will check the `meta.json`'s `"pipeline"` and call `nlp.create_pipe` on each of them. This works fine for the built-ins – but if you're using custom components, you need to tell spaCy how to initialize it. For example:

```python
Language.factories['my_component'] = lambda nlp, **cfg: MyComponent(nlp, **cfg)
```

For more details, check out [this thread over on the Prodigy forum](https://support.prodi.gy/t/adding-custom-attribute-to-doc-having-ner-use-attribute/356/6?u=ines) where I share a full code example of a model's `__init__.py` using a custom `PhraseMatcher` component.

Once the new matcher is implemented in `v2.1.0` (see #1971), I'd love to turn this into a full end-to-end spaCy tutorial.
